### ISSUE TITLE
Execute Courgette having test suite in a FAT JAR

### ISSUE THREAD

**Issue by albertolopez55** [prashant-ramcharan/courgette-jvm#398]
Hi prashant-ramcharan, I’ve been using the courgette extension as my daily basis for a while now, but I still can’t find a way to execute my test suite in a FAT JAR correctly. My biggest problem is that when I try to pass classpath:features as the path to find my feature files like we usually do with cucumber cli.main, it fails to recognize the path when running inside the JAR, can you please provide a solution or include this as part of later releases?

**Comment by prashant-ramcharan** [comment]
Hi @albertolopez55 , can you provide a sample project to demonstrate this?

**Comment by FranciscoAAR** [comment]
Hi @prashant-ramcharan, I'm facing the same issue when trying to execute my test suite from within a JAR. I've created a small project with just one feature and one test scenario in it, the path to the feature files that I'm using is the same as recommended by "java io.cucumber.core.cli.Main" using the "classpath:resources/" format. The problem happens when courgette tries to check feature re-runs, at this point the classpath URI does not contain any path and when the condition tries to retrieve that value fails.

**Feature URI with no path**
<img width="1095" alt="image" src="https://github.com/prashant-ramcharan/courgette-jvm/assets/96503018/aa6d4ff8-7bb2-42c8-b4a3-688381c6623b">

**Attempt to retrieve path from URI**
<img width="1089" alt="image" src="https://github.com/prashant-ramcharan/courgette-jvm/assets/96503018/6d4973e3-3773-41df-9a3e-6dcca5f4e315">

**Below is my runner and main configuration trying to emulate the cucumber CLI feature:**

``` java
package com.test.courgette;

import courgette.api.CourgetteOptions;
import courgette.api.CucumberOptions;
import org.junit.runner.RunWith;
import courgette.api.CourgetteTestOutput;
import courgette.api.CourgetteRunLevel;
import courgette.api.junit.Courgette;
import org.junit.runner.notification.RunNotifier;
import org.junit.runners.model.InitializationError;


@RunWith(Courgette.class)
@CourgetteOptions(
    threads = 1,
    runLevel = CourgetteRunLevel.SCENARIO,
    testOutput = CourgetteTestOutput.CONSOLE,
    rerunFailedScenarios = true,
    rerunAttempts = 1,
    excludeFeatureFromRerun = "",
    excludeTagFromRerun = "",
    reportTitle = "Courgette test execution",
    reportTargetDir = "build/reports/html/chartReport",
    environmentInfo = "",
    cucumberOptions =
    @CucumberOptions(
        features = "src/test/resources/features/",
        glue = "com.test.courgette.steps",
        tags = "@test",
        plugin = {
            "pretty",
            "json:build/reports/json/cucumber.json",
            "html:build/reports/html/cucumber.html",
            "junit:build/reports/cucumber.xml"
        }))
public class TestRunner {

  public static void main(String[] args) throws InitializationError {
    System.setProperty("cucumber.features", "classpath:features/");
    new Courgette(TestRunner.class).run(new RunNotifier());
    System.out.println("\n+++++++++++++++++++++++++\n");
    System.out.println("\n++++EXECUTION FINISHED+++\n");
    System.out.println("\n+++++++++++++++++++++++++\n");
  }
}
```

**Project directory tree**
<img width="554" alt="image" src="https://github.com/prashant-ramcharan/courgette-jvm/assets/96503018/d7ab05aa-06d6-4fe3-abf9-ee55761a8fc0">

I don't know if more problems will arise beyond this point, thanks!

**Comment by prashant-ramcharan** [comment]
Hi @FranciscoAAR Thanks for that.

Do you mind sharing the project so I can debug further, please?

**Comment by FranciscoAAR** [comment]
[Courgette jar runner sample](https://github.com/FranciscoAAR/courgette_jar_runner_sample)

**Comment by FranciscoAAR** [comment]
Hi @prashant-ramcharan - Just following up to see if you need anything else from my side to continue working on this.

**Comment by prashant-ramcharan** [comment]
Hello, I've not had any free time to investigate this further. 

On initial observations though, it's unlikely to work in this way as Courgette does not support this out the box. Courgette does not have a CLI like Cucumber does.

I will need to investigate alternative options to support this.

**Comment by prashant-ramcharan** [comment]
@FranciscoAAR Just had a look into your sample project. 

This seems to work however you should not set the `cucumber.features` to a classpath. 

Courgette and Cucumber should resolve the features using the file system instead.

<img width="1647" alt="Screenshot 2024-06-01 at 10 00 30" src="https://github.com/prashant-ramcharan/courgette-jvm/assets/2563149/1bdde4c0-78d8-4d1a-99f8-aeda2b7431ab">

**Comment by FranciscoAAR** [comment]
@prashant-ramcharan - Thanks for reviewing this. The issue with using the file system is that the feature files need to be outside the JAR for it to find them, thus ignoring the internal resources and failing if not provided. 

<img width="776" alt="image" src="https://github.com/prashant-ramcharan/courgette-jvm/assets/96503018/0376a7f3-9885-4444-b23b-74e9ba5feeec">

Cucumber already supports this out of the box, but I'm noticing that Courgette tries to perform this operation inside _**CourgetteRunnerInfo.checkRerunCondition()**_ :

``` java
Arrays.stream(excludedRerunFeatures)
                .map(String::trim)
                .map(String::toLowerCase)
                .noneMatch(featureName -> feature.getUri().getPath().toLowerCase().contains(featureName))
```

Here `feature.getUri().getPath()` returns _**"null"**_ when the classpath value is used e.g: **_"classpath:features//CourgetteTest.feature"_**

**Comment by prashant-ramcharan** [comment]
Hi @FranciscoAAR ,

I will release an update to the library to support this.

I've tested a version locally with a change to support this. 

Attached jar below:

`````java
java -jar CourgetteJarProject-1.0-SNAPSHOT-all.jar
`````

[CourgetteJarProject-1.0-SNAPSHOT-all.jar.zip](https://github.com/user-attachments/files/15522195/CourgetteJarProject-1.0-SNAPSHOT-all.jar.zip)

Let me know if it runs for you.

**Comment by FranciscoAAR** [comment]
@prashant-ramcharan  - I just ran the jar and it's working as expected!

**Comment by prashant-ramcharan** [comment]
@albertolopez55 @FranciscoAAR This is now released in `6.14.0`

https://github.com/prashant-ramcharan/courgette-jvm/blob/master/CHANGELOG.md#changes-in-version-6140
